### PR TITLE
DOC: Design drafts to assist with next-gen pandas internals discussion

### PR DESCRIPTION
**Pull request by wesm**
Sprinted on my accumulated design ideas. There's more to do, but I'm going to take a break for a bit. 

I published this as a sphinx site temporarily to https://wesm.github.io/pandas2-design/ so it's easier to read. This needs to be a team effort, so depending on where the discussions go perhaps we can build a set of documents that we can make PRs to and discuss specific matters either in GitHub issues or on the mailing list. Let me know what everybody thinks!

### CODE REVIEW

**Comment by gfyoung** [comment]
On the topic of cleaning up API's, IMO there should be a section about the `read_csv` API.  This API is a giant monolith and supporting two implementations doesn't make much sense anymore given that their behaviours and support are inconsistent.  Defining a new, sleeker interface (most likely based in the C engine) would be nice to have.

**Comment by gfyoung** [comment]
One general point of observation so far: this coupling with `numpy` probably will need some more fleshing out, especially when it comes to defining data types in `pandas`.  Finding that good balance between just enough separation from `numpy` to be able to operate largely independently of it while at the same time providing nice inter-op capabilities between `pandas` objects and `numpy` objects is going to be tricky.

**Comment by wesm** [comment]
@gfyoung I would appreciate some help making very specific analysis (with code examples) of potential concerns around insulating users from NumPy-specific implementation details. For example: I do not believe it is pandas's responsibility to maximize its "substitutability" with ndarrays. It's a "nice to have", but the original "Series is an ndarray" design was a mistake in retrospect.

**Comment by wesm** [comment]
I agree that coming up with a plan to improve `read_csv` is a good idea — this is basically orthogonal to the data structure internals. The implementation of `read_csv` will obviously have to change (in a good way, e.g. "unavoidable consolidation" can be avoided in the new design).

**Comment by wesm** [comment]
Also, I would appreciate instead of using the term "data type" rather to use either "physical type" or "logical type".  In NumPy there is no separation of these concepts (except perhaps with `datetime[unit]`) but this is big reason why we are in this mess.

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+
+High-level logical type proposal
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As we have been discussing periodically on the pandas-dev mailing list and
+GitHub, I am proposing that we start to unravel our current mess by defining
+pandas-specific metadata objects that model the current semantics / behavior of
+the project. What does this mean, exactly?
+
+* Each NumPy dtype object will map 1-to-1 to an equivalent ``pandas.DataType``
```
Comment (gfyoung) [inline doc/pandas-2.0/source/internal-architecture.rst]: A layout of a proposed pandas` types class hierarchy would be useful here, especially if we are trying to delineate cleanly between logical and physical types.  `pandas.DataType` confused me a bit especially in light of your section header.

Comment (gfyoung) [inline doc/pandas-2.0/source/internal-architecture.rst]: Also, to what degree would they be equivalent?  Is it just purely semantics?  Or is there some sort sort of physical compatibility as well (i.e. we could "convert" between the two if we so choose)?

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+  pandas objects like Series and DataFrame
+
+* ``numpy.asarray`` will always yield some array, even if it discards metadata
+  or has to create a new array. For example ``asarray`` invoked on
+  ``pandas.Categorical`` yields a reconstructed array (rather than either the
+  categories or codes internal arrays)
+
+* Many NumPy methods designed to work on subclasses (or duck-typed classes) of
+  ``ndarray`` may be used. For example ``numpy.sum`` may be used on a Series
+  even though it does not invoke NumPy's internal C sum algorithm. This means
```
Comment (gfyoung) [inline doc/pandas-2.0/source/internal-architecture.rst]: I am a little wary about this.  I've worked on this compatibility issue from both sides (`numpy` and `pandas`), and in `pandas`, my feeling is that we've had to bend over backwards a bit (this whole directory alone <a href="https://github.com/pydata/pandas/tree/master/pandas/compat/numpy">here</a> speaks for itself) for such compatibility.  Do we really want to commit ourselves to this?

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 1 above)
Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: I can write down a hierarchy. Any `pandas.DataType` is a *logical type* (perhaps we can come up with a better name for this base class?)  — all of the pandas metadata objects are logical, with a clear default mapping onto a physical memory representation. 

For example:

- `pandas.Int64Type` —> `numpy.int64` plus a pandas internal bitmap
- `pandas.Float64Type` —> `numpy.float64`
- `pandas.CategoricalType` —> one of `numpy.int8` to `numpy.int64` depending on the categories
- `pandas.StringType` —> dictionary-encoded UTF8 representation described in the docs

and so on. I'll add it to the document — the main point is that there's a 1-1 mapping from NumPy's physical types onto pandas's logical types, but the mapping from pandas to NumPy may not by 1 to 1, and may not map onto NumPy at all without an explicit lossy cast (instead of the implicit lossy casts happening right now: the example cited in the docs is missing data in integer arrays).

Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: Another thing that isn't explicitly called out in these docs:

- The pandas logical types are only metadata

- NumPy's physical dtypes are metadata plus a vtable of C functions (the `f` attribute on the `PyArray_Descr` object) that implement algorithms on those types

The resolution of a particular function will depend on the *actual* physical memory representation of the data in pandas.

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+  copy-on-write) which may yield many benefits. I will write a dedicated
+  section about this.
+
+Note that neither of these points implies that we are trying to use NumPy
+less. We already have large amounts of code that implement algorithms similar
+to those found in NumPy (e.g. ``pandas.unique`` or the implementation of
+``Series.sum``), but taking into account pandas's missing data representation,
+etc. Internally, we can use NumPy when its computational semantics match those
+we've chosen for pandas, and elsewhere we can invoke pandas-specific code.
+
```
Comment (gfyoung) [inline doc/pandas-2.0/source/internal-architecture.rst]: I very much agree with this and think it is a crucial point that could be emphasised more in the docs.  I suspect that this is a tripping point for many users regarding the application of `numpy` functionality on `pandas` objects.

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 1 above)
Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: On your other question re: physical compatibility:

https://wesm.github.io/pandas2-design/internal-architecture.html#preserving-numpy-interoperability

- If the mapping between a pandas logical type and a physical NumPy type is 1 to 1 (for example: `float64`), then you can convert to ndarray and back without copying data

- Some types will not be representable as `ndarray` without a lossy conversion. For example: a pandas integer array containing nulls can be converted to ndarray but this conversion will be "lossy" (unless you use `numpy.ma`). For example, you may choose the current behavior which is conversion to `float64` (this is already a one-way trip for data — values exceeding 2^53 get destroyed by this)

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 2 above)
Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: I deliberately chose pretty hedgy language here. "Many methods ... may be used" instead of "All methods ... can be used". Maybe we should say that we're going to back away from this here?

Maintaining an explicit API contract that you can pass a Series or DataFrame wherever you might otherwise pass an ndarray seems like a bad idea to me, and is likely to be a continuous source of bugs and maintainability problems. I know you've worked a lot of this and very honestly I cringed when I saw the patches come through — I would rather not reinforce this interchangeability in users' minds. I agree that accommodating NumPy functions where it does no harm to pandas is okay, but I don't think it's our responsibility to do so.

But honestly: the average pandas user is, in my anecdotal experience, more familiar with pandas's features than NumPy's features, so the audience of people for whom "semantic parity" with NumPy is important is probably growing smaller over time. I would rather focus on creating a consistent and self-contained user experience in pandas.

Comment (gfyoung) [inline doc/pandas-2.0/source/internal-architecture.rst]: I think toning down the promise of inter-op functionality with `numpy` would be a good idea, and yes, I agree: consistent with decoupling from `numpy` is to create a more self-contained user experience in `pandas`.

File: doc/pandas-2.0/source/goals.rst [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  reflect his opinions for the time being. Many things may change as we discuss
+  and work to reach a consensus about the path forward.
+
+The pandas codebase is now over 8 years old, having grown to over 200,000 lines
+of code from its original ~10,000 LOC in the original 0.1 open source release
+in January 2010.
+
+At a high level, the "pandas 2.0" effort is based on a number of observations:
+
+* The pandas 0.x series of releases have consisted with huge amounts of
```
Comment (TomAugspurger) [inline doc/pandas-2.0/source/goals.rst]: This first point muddles the waters a bit (for me); likewise with the point starting on line 79. " **Removal of deprecated / underutilized functionality**". 

Further down you seem to split between

1. API changes that are the result of pandas 0.X implementation details (e.g. fixing integer-NA), and
2. API changes that would be because the original idea may be flawed / out of scope (`.ix` `plotting`).

Does it make sense to limit this document *entirely* to the internals refactoring, and only talk about  API changes of the first kind?

File: doc/pandas-2.0/source/goals.rst [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+* Rough edges in pandas's implementation (e.g. its handling of missing data
+  across data types) are being exposed to users.
+
+These documents are largely concerned with pandas's internal design, which is
+mostly invisible to average users. Advanced users of pandas are generally
+familiar with some of these internal details, particular around performance and
+memory use, and so the degree to which users are impacted will vary quite a
+lot.
+
+Key areas of work
```
Comment (TomAugspurger) [inline doc/pandas-2.0/source/goals.rst]: I just noticed that the "Goals" page doesn't have a section where we clearly state the actual goals. The "Key areas of work" is probably the closest, but I wouldn't say "Decoupling form NumPy" is itself a goal. Rather, the goal of separating logical-types vs. physical storage requires a physical decoupling. Perhaps trim other sections, and add a section like:

Goals
====

- Integer-NA / bool; More consistent NA handling in general (achieved by the logical type vs. physical storage decoupling).
- clearer semantics around Categorical, DatetimeTZ, and future "extension dtypes"(like pd.String; again, from the decoupling)
- Expose a `libpandas` C/C++ API for other libraries (a scikit-learn dev said in the past that they couldn't depend on pandas in part because there wasn't a C API they could use; I'll try to find that issue)
- User control / visibility of memory / Copy on write
- Improved performance
- Removal of deprecated functionality. (but maybe keep in separate discussion)
- Analytical expression VM (was in your [original mail](https://mail.python.org/pipermail/pandas-dev/2016-July/000512.html), but I didn't see it here; maybe out of scope for now.)

File: doc/pandas-2.0/source/goals.rst (hunk 4 above)
Comment (wesm) [inline doc/pandas-2.0/source/goals.rst]: Good points. I mainly fixated on `.ix` because the `.loc` and `.iloc` indexing operators will probably need to be reimplemented as part of this internals overhaul, and having to drag along things like `.ix` would add implementation burden for unclear benefit. I agree talking about other refactoring / cleaning is a distraction. Will make some amendments to make this more clear.

The other intent of this first point was that the iterative / agile development style of the project (from its early days until now) has made it difficult to consider large/invasive changes to the internals, and after so much time we are due to seriously contemplate what's working well and what's not working well.

File: doc/pandas-2.0/source/goals.rst (hunk 5 above)
Comment (wesm) [inline doc/pandas-2.0/source/goals.rst]: Agreed -- staying focused on the goals (without diving into the possible solutions) will help steer the discussion. If something is controversial, it helps to return to the goal since tradeoffs / API semantics changes will have to be made in the interest of achieving some of the goals. There may also be other approaches to achieving the results that we haven't thought of.

**Comment by chris-b1** [comment]
Thanks for putting this together!

A topic that is somewhat orthogonal to the internals, but one thing I've been thinking about in the context of pandas 2.0, is interaction with JIT compilers - mostly for UDFs.  I'm using `numba` as a frame of reference, but thinking generally.

I came to pandas from SAS - there is _plenty_ to dislike about it - but one thing I actually do miss is a feeling of safety, that if you can write it, it will basically be fast.  Of course with pandas/numpy you have to much more defensive, and sometimes try really hard to vectorize - just as an example I remember, this [SO answer](http://stackoverflow.com/a/32524908/3657742) from @unutbu is really clever, but I would have never come up with it.

I think this need to be vectorized adds to the pandas api  size, and I also think adds significantly to the learning curve.  I'm sure like many others, I've taken a hard look at Julia - it wasn't there yet (and maybe it won't ever be?) but there is a big allure to that model.  As a user, I'd love to be able to just write this, and just have it be fast.
```python
df.groupby('key').transform(lambda x: (x - x.mean()) / x.std())
```

This may be more of a stretch, but as a contributor who's not a C++ wizard, it would also be nice if some portion of pandas was implementable at a JIT-able level, ala @shoyer's [numbagg](https://github.com/shoyer/numbagg).  I get this may not possible - you probably couldn't even make `numba` a required dep.

So I was just curious if you had any thoughts / roadmap ideas for this topic?  A C/C++ API would open the door a lot of the way, but I think there may be value in tighter integration?

File: doc/pandas-2.0/source/strings.rst [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    each value.
+
+Possible solution: new non-NumPy string memory layout
+=====================================================
+
+My proposed solution to the string conundrum is the following:
+
+* Create a custom string array container type suitable for use in a
+  ``pandas.Array``, and a ``pandas.string`` logical data type.
+* Require that all strings be encoded as UTF-8.
```
Comment (MaximilianR) [inline doc/pandas-2.0/source/strings.rst]: What are the benefits of UTF-8 vs Unicode? It is just space? If encoding categorically, would the space become less of an issue? 
Would Unicode allay some of the issues below?
And provide easier compatibility with python3?

Comment (shoyer) [inline doc/pandas-2.0/source/strings.rst]: My understanding is that "Unicode" is a generic term, versus UTF-8 refers to a specific encoding. Python's Unicode type uses UTF-8 internally, but also some other metadata (type, reference counter plus reference to the data -- this is the 24 bytes of PyObject overhead) that means creating it requires memory allocation.

File: doc/pandas-2.0/source/removals.rst [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+* **Monotonicity**: for comparable data (e.g. numbers), is the data sorted /
+  strictly increasing? In time series, this permits sorting steps to be
+  skipped.
+
+* **Null count**: for data not containing any nulls, the null handling path in
+  some algorithms can be skipped entirely
+
+
+
+Strided arrays: more trouble than they are worth?
```
Comment (shoyer) [inline doc/pandas-2.0/source/removals.rst]: Totally agreed.

File: doc/pandas-2.0/source/strings.rst (hunk 6 above)
Comment (MaximilianR) [inline doc/pandas-2.0/source/strings.rst]: I see. I didn't know python's internal representation was variable length. Thanks

Comment (wesm) [inline doc/pandas-2.0/source/strings.rst]: See https://www.python.org/dev/peps/pep-0393/

Comment (shoyer) [inline doc/pandas-2.0/source/strings.rst]: OK, so it's a bit more overhead than I thought :).
```
>>> sys.getsizeof(unicode())
52
```

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 1 above)
Comment (jreback) [inline doc/pandas-2.0/source/internal-architecture.rst]: I would add:

- ``PeriodDtype`` [here](https://github.com/pydata/pandas/pull/13941)
- [pd.String](https://github.com/pydata/pandas/issues/8640) (the link)
- future possibilities, e.g. ``Interval`` [here](https://github.com/pydata/pandas/issues/8625) (which ``Period`` is really just a sub-type)

File: doc/pandas-2.0/source/strings.rst (hunk 6 above)
Comment (jreback) [inline doc/pandas-2.0/source/strings.rst]: But in context its not so bad (this is PY2).

```
In [3]: sys.getsizeof(str())
Out[3]: 33

In [4]: sys.getsizeof(unicode())
Out[4]: 50
```

of course a buffered string pointer is 1

**Comment by wesm** [comment]
@chris-b1 several things on the "faster groupby" front:

- if the inner loops of group-wise operations are happening in compiled code with limited contact wtih the Python interpreter, then the microperformance of each operation will be much better

- hypothetically (with much effort) one could create a useful enough C API such that the groupby operators could accept a ctypes function pointer -- this is basically (IIUC) how numba is implementing custom ufuncs. obviously the low-level bits of numerical arrays are simpler than pandas objects but the same general approach would be valid.

File: doc/pandas-2.0/source/goals.rst [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+* **Improving performance and system utilization**: We aim to improve both the
+  micro (operations that take < 1 ms) and macro (all other operations)
+  performance of pandas across the board. As part of this, we aim to make it
+  easier for pandas's core developers to leverage multicore systems to
+  accelerate computations (without running into any of Python's well-known
+  concurrency limitations)
+* **Removal of deprecated / underutilized functionality**: As the Python data
+  ecosystem has grown, a number of areas of pandas (e.g. plotting and datasets
+  with more than 2 dimensions) may be better served by other open source
+  projects. Also, functionality that has been explicitly deprecated or
```
Comment (jreback) [inline doc/pandas-2.0/source/goals.rst]: some should add: enforce groupby immutability. This has caused so many random issues with the current inference (e.g. someone mutates *inside* the udf). a COW on pandas objects could detect this much more reliably that the current way (e.g. seeing if the index was copied). Should simply disallow this.

File: doc/pandas-2.0/source/goals.rst [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+  discouraged from use (like the ``.ix`` indexing operator) would ideally be
+  removed.
```
Comment (jreback) [inline doc/pandas-2.0/source/goals.rst]: we should discuss the status of sparse. @sinhrks has basically made this pretty well supported recently. Should it be a dtype though? or just an impl detail, e.g. ``int64[sparse]`` does have some nice appeal.

Comment (wesm) [inline doc/pandas-2.0/source/goals.rst]: Seems reasonable to push down the sparse implementation into libpandas. You could hypothetically eliminate the Sparse* classes altogether.

Comment (shoyer) [inline doc/pandas-2.0/source/goals.rst]: Yes, I would either move sparse into libpandas (if it can be done cleanly) or drop it. But I would hesitate to hold up the rest of pandas 2.0 for the sake of sparse, which IMO is pretty niche.

**Comment by jreback** [comment]
so we have tons of issues related to this, see [here](https://github.com/pydata/pandas/issues?q=is%3Aopen+is%3Aissue+label%3AInternals).

I will post a mini-list of issue that I am closing.

File: doc/pandas-2.0/source/removals.rst [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+Other ideas
+===========
+
+Here's a collection of other miscellaneous ideas that don't necessarily fit
+elsewhere in these documents.
+
+Column statistics
+~~~~~~~~~~~~~~~~~
+
+In quite a few pandas algorithms, there are characteristics of the data that
```
Comment (shoyer) [inline doc/pandas-2.0/source/removals.rst]: We do currently calculate on demand and cache some of these properties (at least monotonicity) on the `pandas.Index`, which we can do because it's immutable. I don't see any difficulty extending that to Series or DataFrame (pandas 2.0 or not) as long as we invalidate on mutation.

Comment (wesm) [inline doc/pandas-2.0/source/removals.rst]: Yeah, we would need some kind of a "dirty" flag for the statistics so that any mutation methods sets `dirty = true`

File: doc/pandas-2.0/source/goals.rst (hunk 9 above)
Comment (sinhrks) [inline doc/pandas-2.0/source/goals.rst]: I expect sparse can be implemented in ``libpandas`` in a similar way to ``Categorical`` in the sense that it's a mapper from exposed key to actual value. I think sparse is useful if we also targets little larger data.

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 11]
```diff
@@ -0,0 +1,714 @@
...
+at all, and why we should consider removing it.
+
+What is ``BlockManager`` and why does it exist?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The reason that ``BlockManager`` exists at all goes back to some ancient pandas
+history. Originally, the data in ``pandas.DataFrame`` was stored in a Python
+``dict`` object. If you pull up pandas 0.1 or 0.2, you will see this.
+
+Since the business logic of pandas's internals was originally implemented in
```
Comment (jreback) [inline doc/pandas-2.0/source/internal-architecture.rst]: [here](http://nbviewer.jupyter.org/github/jreback/PandasTalks/blob/master/performance/may_2016/1.%20storage.ipynb) is a notebook that I show to explain *why* block manager originally exists IMHO. The implication is that construction from an already formed 2-d ndarray like (with strided memory layout), is going to be substantially more expensive going forward (as it will require splitting into contiguous blocks). Of course this is much less common that in 2011 (and much more common is already contiguous 1D data).

just raising this for awareness.

File: doc/pandas-2.0/source/removals.rst (hunk 10 above)
Comment (jreback) [inline doc/pandas-2.0/source/removals.rst]: uniqueness is important in some columns. potentially compression ability (e.g. imagine backing by a chunked-compressed store).

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 11 above)
Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: Thanks for creating that, I will link to in the document!

I will also definitely mention here that zero-copy creation from 2D ndarrays was a primary motivation (I had dealt with a number of applications in the past where this was a bottleneck). I suspect we'll be able to make the constructor significantly faster in libpandas -- maybe with some tricks, like processing L2-ish cache size contiguous chunks of the matrix at a time to reduce the perf hit of strided memory access.

File: doc/pandas-2.0/source/removals.rst (hunk 10 above)
Comment (wesm) [inline doc/pandas-2.0/source/removals.rst]: yeah, I'll add that here (since we're already computing this in indexes) -- more expensive than the others since you have to push the data through hash table

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 12]
```diff
@@ -0,0 +1,714 @@
...
+    ``ndarray`` and to assign these values to a C type (or, in the case, of
+    structured dtypes, assign to a packed C struct).
+
+  * This may or may not include a physical representation of NULL or missing
+    data (for example: nullable float64 might be a physical type indicating a
+    normal float64 array along with a bitmap of null/not-null indicators).
+
+* **Logical data type**: metadata which describes the semantic content of a
+  single value in an array or other collection of values. Depending on the
+  logical type, it may map 1-to-1 to a physical type or not at all. Here are
```
Comment (jreback) [inline doc/pandas-2.0/source/internal-architecture.rst]: there is an argument that the spelling of the logical types should be more in line with Python (that even numpy) is: https://www.python.org/dev/peps/pep-0484/

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 13]
```diff
@@ -0,0 +1,714 @@
...
+  some examples:
+
+  * The ``double`` or ``float64`` type may be viewed both as a logical type as
+    well as a physical type (a 1-to-1 correspondence).
+
+  * pandas's ``category`` dtype contains its own auxiliary array of category
```
Comment (jreback) [inline doc/pandas-2.0/source/internal-architecture.rst]: might be some nice semantics from libndt (I think it's separate from DyND), http://www.bytereef.org/1112730/datashape/build/html/datashape.html#type-kinds

Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: My understanding has been that these are all indicative of physical memory representation (a superset of the NumPy physical type grammer supporting variable length types, etc.)

Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: Let me rephrase that -- what I meant by "physical memory representation" was describing the "shape" or "physical structure" of an array cell. The actual memory buffers in an array would need to be defined in an array library implementation like libdynd

Comment (jreback) [inline doc/pandas-2.0/source/internal-architecture.rst]: not that I know a lot about this 

but it seems to me that logical as a super set +
combination of physical types makes sense

though I wonder why the distinction is not made in most type systems?

**Comment by shoyer** [comment]
As for C vs C++, one factor that I find convenient about C++ is that you have the STL for standard data structures builtin. NumPy actually uses Python's C API when it needs a dict or set internally, for example, which is kind of insane.

File: doc/pandas-2.0/source/internal-architecture.rst (hunk 13 above)
Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: As one example, consider the "nullable int64" logical type (in pandas all types are nullable, but suppose you had the same distinction as in a database like PostgreSQL). You can talk about this type without an explicit mapping to a physical representation. You could even support multiple implementations, e.g.:

- 8 bytes plus a null bit (or byte)
- 8 bytes, with a sentinel value

As another example, see the Parquet file format docs: https://github.com/apache/parquet-format/blob/master/LogicalTypes.md

> "Logical types are used to extend the types that parquet can be used to store, by specifying how the primitive types should be interpreted."

**Comment by wesm** [comment]
Note that the STL containers (like `std::unordered_map` for hash tables) have pretty acceptable performance:

http://incise.org/hash-table-benchmarks.html

We should obviously do our own investigations, but we always have the option to pull in 3rd party libraries (we might decide to continue using klib or decommission depending on benchmarks).

File: doc/pandas-2.0/source/goals.rst [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+Possible changes or improvements to pandas's internals fall into a number of
+different buckets to be explored in great detail:
+
+* **Decoupling from NumPy while preserving interoperability**: by eliminating
+  the presumption that pandas objects internally must contain data stored in
+  NumPy ``ndarray`` objects, we will be able to bring more consistency to
+  pandas's semantics and enable the core developers to extend pandas more
+  cleanly with new data types, data structures, and computational semantics.
+* **Exposing a pandas Cython and/or C/C++ API to other Python library
+  developers**: the internals of Series and DataFrame are only weakly
```
Comment (jreback) [inline doc/pandas-2.0/source/goals.rst]: I think it's prob worth some thought how to make a clean API between the storage back end, pandas structures and the dispatch (compute) machinery

eg say we wanted to swap in something like: https://github.com/alimanfoo/zarr

this actually in theory should be easy as it has numpy storage / indexing semantics 

in a similar way would be nice to define an API to a compute engine (like numexpr or numba)

mainly need a consistent way of handling dtype conversions (this of course is very similar to the problem arrow is designed to solve) except here we could use numpy as the intermediary

Comment (wesm) [inline doc/pandas-2.0/source/goals.rst]: A probable pre-requisite for 3rd-party data structures would be having a C or C++ API (not sure if this one does, but in general), unless they get coerced to NumPy arrays whenever you need to do anything that falls outside their pure Python API.

**Comment by wesm** [comment]
I'm going to let discussions collect here for a bit, then incorporate feedback into the documents. 

It would be useful to start a separate document that is changes/improvements/refactorings that we would like to do in pandas 2.0 that do not involve changes to the internals, how do you want to manage that?

**Comment by shoyer** [comment]
One thing that surfaced recently in #13395 that I would like consider for pandas 2.0 is to expose the contiguous one dimensional arrays that store the values of DataFrame colums as part of the public API. These would be similar to the existing `Categorical` type, but done in more systematic manner and returned by all pandas objects when you call `Series.values`.

This would be complementary to existing scalar types like `Period`/`Datetime`/`Timestamp`. If the existing scalar type has the name `pandas.Scalar`, I would call the array version `pandas.ScalarArray`, inheriting from `pandas.BaseArray`.

Having such an API would be useful for third party libraries that integrate with pandas in a deep way (such as seaborn, sklearn, statsmodels and xarray), because it doesn't always make sense to work with labeled values.

**Comment by wesm** [comment]
@shoyer I'm totally with you here. This isn't explicitly called out in these docs but that was kind of what I intended here:

https://wesm.github.io/pandas2-design/internal-architecture.html#pandas-array-types

So basically every logical type (whether Int32Array, StringArray, CategoricalArray, etc.) would have a corresponding public Python API. These would also have `__array__` for NumPy interoperability as able.

File: doc/pandas-2.0/source/internal-architecture.rst [hunk 15]
```diff
@@ -0,0 +1,714 @@
...
+  result in better CPU cache utilization (fewer main-memory reads of the
+  bitmap).
+* Hardware and SIMD "popcount" intrinsics (which can operate on 64-128 bits at
+  a time) can be used to count bits and skip null-handling on segments of data
+  containing no nulls.
+
+Notably, this is the way that PostgreSQL handles null values. For example, we
+might have:
+
+.. code-block::
```
Comment (chrisaycock) [inline doc/pandas-2.0/source/internal-architecture.rst]: This code block is not showing up in the generated HTML document. Is a language required?

Comment (wesm) [inline doc/pandas-2.0/source/internal-architecture.rst]: Yep, apparently code-block does not default even to plain text if you don't indicate the language. I'll fix soon on my pass through to incorporate all the feedback here

**Comment by jreback** [comment]
I realigned the milestones with some 'approx' dates.

File: doc/pandas-2.0/source/goals.rst (hunk 4 above)
Comment (jorisvandenbossche) [inline doc/pandas-2.0/source/goals.rst]: +1 on making this distinction. We can also start drafting documents that layout other API changes (not related to internal refactoring) and put those in the same directory (so the 'goals and motivations' can touch both aspects), but in separate PRs

**Comment by jorisvandenbossche** [comment]
@wesm Thanks for the extensive explanations!

Two small things I wondered:

- There has been discussion about the the datetime resolution (and resulting span) which is limited. I suppose with the better logical dtype system, it should be possible to tackle this? But maybe it is worth mentioning it somewhere?

- Also regarding the dtype system, to what extent would it be possible to make defining custom dtypes externally work? (other libraries that provide a dtype that can be used in pandas dataframes)

**Comment by jorisvandenbossche** [comment]
> It would be useful to start a separate document that is changes/improvements/refactorings that we would like to do in pandas 2.0 that do not involve changes to the internals (like, read_csv was brought up), how do you want to manage that? 

Agreed. 
Would it be worth to create a separate repo where those documents can live, together with the ones of this PR? (a pandas-design, pandas-discussion, pandas-dev, ...) This way, we can also use the issues/PRs over there to discuss the several topics without adding to the already huge number of open issues in the pandas repo, keeping it a bit separated. 
But this has the disadvantage that when we start with actual code PRs, it will again be in the pandas repo.

Otherwise, I would just add documents describing those possible changes to this same ``pandas-2.0`` directory.

**Comment by wesm** [comment]
@jorisvandenbossche I'd be happy to create a pandas-design repo — this has the added benefit of not bloating the main pandas git history if we check in images or other assets. Otherwise we can keep working in this doc/pandas-2.0 directory and indicate which docs are "internals-related" and which are "stuff that depends on the internals" (like the CSV reader, indexing, etc.)

**Comment by wesm** [comment]
I'm going to start incorporating the feedback into this document. I'll write a follow up PR for the Copy-On-Write discussion. I'll indicate in the document pages that are Internals related and those that are separate from Internals and maybe we can make more PRs digging into some of the other refactoring / cleanup we want to do. 

Let's start discussing on the mailing list how we intend to proceed with two branches of pandas? I would suggest we should keep the pandas 2.0 branch "rebaseable" for as long as possible, but at some point there will be some necessary divergence.

**Comment by jorisvandenbossche** [comment]
What do others think of putting those documents and PRs in a separate repo?

**Comment by wesm** [comment]
I'm +0 on a separate repo (pydata/pandas-design or something), if only to make the discussion issues more accessible to outsiders (vs. having to seek them out in the main pandas issue tracker)

**Comment by wesm** [comment]
I'll go ahead and do that. We can always move the docs back here if the process is not working well. 

I also would like more people to actively watch the design discussion. Having a dedicated repo will make the GitHub emails separate from the usual pandas firehose.

**Comment by wesm** [comment]
I've addressed many of the comments in here in pydata/pandas-design#1 -- let's move the discussion there and see how it goes. I still need to fix up the github permissions so that all who have push rights here can push there
